runtime/js/40_testing.js

```diff
@@ -815,21 +839,56 @@
     });
 
     try {
-      const warmupIterations = bench.warmupIterations;
-      step.warmup = true;
+      // Check the return type of the bench so we can use specialized
+      // function to run it. One might be tempted to just always await
+      // `bench.fn`, but that adds a lot of overhead to synchronous
+      // benches.
+      //
+      // First we'll wrap the bench function with sanitizers and permissions
+      // handler and probe if it async or sync bench.
+      //
+      // Only then, we'll wrap the bench case again with iteration reporter
+      // and override `bench.fn`.
+      let benchFn = wrapBenchFnWithSanitizers(
+        bench.fn,
+        bench,
+      );
 
-      for (let i = 0; i < warmupIterations; i++) {
-        await bench.fn(step);
+      if (bench.permissions) {
+        benchFn = withPermissions(
+          benchFn,
+          bench.permissions,
+        );
       }
 
-      const iterations = bench.n;
-      step.warmup = false;
+      const result = benchFn(step);
 
-      for (let i = 0; i < iterations; i++) {
-        await bench.fn(step);
+      if (typeof result.then !== "undefined") {
```

Comment:
Perhaps check for "function" type? Although rare, it is theoretically possible that a sync function would return an object like
```ts
const result = {
  then: 3,
};
```

This is not promise-like.

A further check could be done to make sure that the `.then` function has a length of at least one:

```ts
const result = {
  then: () => 3,
};
```

This isn't a valid promise-like object either, doing `await` on that will never resolve.

So a full check could be something like this:

```ts
// Quick-check for normal Promises
if (result instanceof Promise || typeof result === "object" && result !== null && typeof result.then === "function" && result.then.length >= 1) {
  // ...
}
```

One could also consider doing some sort of a test-by-fire on a potential promise-like:

```ts
async function testPromiseLike(maybePromiseLike: unknown, bench) {
  if (typeof maybePromiseLike !== "object" || maybePromiseLike === null) {
    return false;
  } else if (typeof maybePromiseLike.then !== "function" || maybePromiseLike.then.length === 0) {
    return false;
  }
  // Looks like a promise, now lets see if it walks like a promise
  return await Promise.race([
    maybePromiseLike.then(() => true, () => true), // Return or throw, it doesn't matter
    new Promise((res) => setTimeout(() => res(false), bench.timeout)), // Does not resolve within the bench timeout, this does not seem like a real promise
  ]);
}
```

Now if the return value looks and walks like a Promise, it'll get used like a promise. Otherwise it'll be treated as a sync function.